### PR TITLE
`requires_grad` *should* be carried through expressions

### LINKED ISSUE #370: Possible bug with `requires_grad` attribute being changed following operations
I have observed a case where the `requires_grad` attribute for a tensor is flipped from `.false.` to `.true.` as a result of assigning it based on a calculation involving other tensors.

I have created a MWE to showcase this and will open a PR to demonstrate.

The key lines are that:

```fortran
    output_vec = input_vec * scaling_tensor
    grad_reqd = loss%requires_grad()
    write(*,*) "loss%requires_grad = ", grad_reqd

    loss = ((output_vec - target_vec) ** 2)
    grad_reqd = loss%requires_grad()
    write(*,*) "loss%requires_grad = ", grad_reqd
```
cause `requires_grad` for `loss` to go from `.false.` to `.true.`.

The only tensor in the entire example that has `requires_grad=.true.` set is `scaling tensor`.

This is an issue because it prevents values being assigned to `loss` in subsequent calculations.

### PR DESCRIPTION
**Pull request by jwallwork23**
Closes #370.

It turns out I was wrong in #364. Some simple Python experimentation shows that we *should* be carrying `requires_grad` over through expressions.
```
>>> a = torch.Tensor([1])
>>> a.requires_grad = True
>>> a
tensor([1.], requires_grad=True)
>>> b = torch.Tensor([2])
>>> c = a + b
>>> c
tensor([3.], grad_fn=<AddBackward0>)
>>> c.requires_grad
True
```
This PR reverts the changes from #364 and drops the related check that was causing issues.

Further, we should actually be overwriting with `requires_grad=False` when appropriate, too:
```
>>> import torch
>>> a = torch.Tensor([1])
>>> a.requires_grad=True
>>> a
tensor([1.], requires_grad=True)
>>> b = torch.Tensor([1])
>>> b
tensor([1.])
>>> a = b + b
>>> a
tensor([2.])
```

### CODE REVIEW

**Comment by jwallwork23** [comment]
Added a load of testing. There are some cases that seem to be inconsistent with what I'd expect from PyTorch. (See the `NOTE`s in the unit tests.)

File: test/unit/test_tensor_autograd.pf [hunk 1]
```diff
@@ -28,13 +28,128 @@ module test_tensor_autograd
...
+  @testParameter
+  type, extends(AbstractTestParameter) :: TestParametersType
+    logical :: switch
+    integer :: op
+  contains
+    procedure :: toString
+  end type TestParametersType
+
+  ! Typedef for a test case with a particular set of parameters
+  @testCase(constructor=test_case_ctor)
```
Comment (jatkinson1000) [inline test/unit/test_tensor_autograd.pf:43]:
```suggestion
  @testCase(constructor=test_case_constructor)
```
I spent a while trying to work out what `ctor` was. I would suggest perhaps naming this explicitly (doesn't feel overly verbose) here and elsewhere where used.

File: test/unit/test_tensor_autograd.pf [hunk 2]
```diff
@@ -28,13 +28,128 @@ module test_tensor_autograd
...
+  end function get_parameters_short
+
+  ! Constructor for the test case type
+  function test_case_ctor(param)
+    type(TestCaseType) :: test_case_ctor
+    type(TestParametersType), intent(in) :: param
+    test_case_ctor%param = param
+  end function test_case_ctor
+
+  ! Function for representing a parameter set as a string
```
Comment (jatkinson1000) [inline test/unit/test_tensor_autograd.pf:113]: I'm not wild about this large if-else, but I have no better suggestions at the moment and it's not production code.
I can see why the feedback to stdout is useful.

File: test/unit/test_tensor_autograd.pf [hunk 3]
```diff
@@ -119,5 +235,201 @@ contains
...
+      tensor3 = tensor1 + tensor2
+    else if (this%param%op == 2) then
+      tensor3 = tensor1 - tensor2
+    else if (this%param%op == 3) then
+      tensor3 = tensor1 * tensor2
+    else if (this%param%op == 4) then
+      tensor3 = tensor1 / tensor2
+    else
+      write(*,*) "Error :: invalid operator code"
+    end if
```
Comment (jatkinson1000) [inline test/unit/test_tensor_autograd.pf:303]: Note to self, I might steal this approach in #320 as a way of running tests over all optimizer types.

**Review by jatkinson1000 — APPROVED** [review]
Thanks for looking at this @jwallwork23 

Code wise it looks good, I just had a couple of minor suggestions.

Only other thought is that I wonder if we should explicitly document the requires grad behaviour anywhere? Whilst it is true that it 'matches that expected from PyTorch' it caught both of us out... ;)

File: test/unit/test_tensor_autograd.pf (hunk 2 above)
Comment (jwallwork23) [reply test/unit/test_tensor_autograd.pf:113]: Ditto

File: test/unit/test_tensor_autograd.pf (hunk 1 above)
Comment (jwallwork23) [reply test/unit/test_tensor_autograd.pf]: It's a fairly standard abbreviation in object-oriented programming, but sure. Done in 3eccb29f164a4a03099c484c7f6c83f20e1ea58a.
